src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -199,6 +199,18 @@ def upgrade_start(self, image: str, version: str) -> str:
                 self._save_upgrade_state()
                 return 'Resumed upgrade to %s' % self.target_image
             return 'Upgrade to %s in progress' % self.target_image
+
+        running_mgr_count = len([daemon for daemon in self.mgr.cache.get_daemons_by_type(
+            'mgr') if daemon.status == DaemonDescriptionStatus.running])
+
+        running_mon_count = len([daemon for daemon in self.mgr.cache.get_daemons_by_type(
+            'mon') if daemon.status == DaemonDescriptionStatus.running])
+
+        if running_mgr_count < 2:
+            raise OrchestratorError('Need at least 2 running mgr daemons for upgrade')
+        if running_mon_count < 3:
+            raise OrchestratorError('Need at least 3 running mon daemons for upgrade')
```

Comment:
Yeah, I think we have a couple options.

1. mon restarts are fast, so we could forgo the ok-to-stop in all cases.
2. if we are in a non-ok-to-stop state at the start of upgrade, we set a flag, and forgo mon ok-to-stop; otherwise, we behave like we currently do.
3. same as 2, but we warn, and ask the user to pass --force to proceed with upgrade even with the risk of a transient loss of mon availability.

I lean toward 2 for simplicity, and because for most clusters, the difference between a mon election and a mon restart is not noticeable (and mon outages in general do not disrupt IO)

For managers:

1. proceed with a check like this PR makes sense since we cannot currently proceed with <2
2. make it so that ceph-mgr can redeploy itself in place

I'm inclined to go with 1 until we demonstrate that 2 works